src/core/i18n/translate.js

```diff
@@ -49,6 +44,12 @@ export default function translate(options = {}) {
 
     Translate.displayName = `Translate[${getDisplayName(WrappedComponent)}]`;
 
-    return Translate;
+    function mapStateToProps({ api, i18n }) {
+      return {
+        i18n: makeI18n(i18n, api.lang),
```

Comment:
Making `translate()` a pure component might fix this if we move the call into the component itself but I am not sure. It is quite possible that this would break the underlying child components however since I the redux store is in the context and it is probably being updated to propagate changes. There might be some extra magic going on to keep redux from breaking but I'm not sure.

Reselect definitely seems like the right approach to me.